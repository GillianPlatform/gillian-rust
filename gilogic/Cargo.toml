[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }