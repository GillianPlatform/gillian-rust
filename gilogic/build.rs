fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
