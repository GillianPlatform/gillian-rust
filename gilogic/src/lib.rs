#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
