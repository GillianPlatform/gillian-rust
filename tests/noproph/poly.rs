//@check-pass
extern crate gilogic;

use gilogic::{macros::show_safety, Ownable};

#[show_safety]
fn test<T: Ownable>(x: Box<T>) -> T {
    return *x;
}

#[show_safety]
fn test2() -> i32 {
    test(Box::new(12))
}
