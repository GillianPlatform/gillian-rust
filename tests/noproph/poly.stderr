[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"std"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
building predctx DefId(0:7 ~ poly[df2d]::test_spec_e154afdf_03f0_4c3c_88ac_95856e80d60c) [T/#0]
[TIMESTAMP WARN  lib_rtg::codegen::block] Not handling drop properly yet: _1
building predctx DefId(0:7 ~ poly[df2d]::test_spec_e154afdf_03f0_4c3c_88ac_95856e80d60c) [T/#0]
building predctx DefId(0:13 ~ poly[df2d]::test2_spec_93c8120a_9671_4758_b740_25bd742f60ca) []
building predctx DefId(0:13 ~ poly[df2d]::test2_spec_93c8120a_9671_4758_b740_25bd742f60ca) []
building predctx DefId(20:235 ~ gilogic[8e9c]::ownable::{impl#3}::own) [T/#0]
building predctx DefId(20:189 ~ gilogic[8e9c]::ownable::Ownable::own) [T/#0]
building predctx DefId(20:219 ~ gilogic[8e9c]::ownable::{impl#13}::own) []
building predctx DefId(0:7 ~ poly[df2d]::test_spec_e154afdf_03f0_4c3c_88ac_95856e80d60c) [i32]
building predctx DefId(20:235 ~ gilogic[8e9c]::ownable::{impl#3}::own) [i32]
warning: function `test` is never used
 --> ../tests/noproph/poly.rs:7:4
  |
7 | fn test<T: Ownable>(x: Box<T>) -> T {
  |    ^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: function `test2` is never used
  --> ../tests/noproph/poly.rs:12:4
   |
12 | fn test2() -> i32 {
   |    ^^^^^

warning: 2 warnings emitted

[TIMESTAMP DEBUG rust_to_gil] Correct!
